optimum/neuron/modeling.py

```diff
@@ -653,3 +660,286 @@ def prepare_inputs_for_generation(self, input_ids: torch.Tensor, **kwargs) -> Di
     def can_generate(self) -> bool:
         """Returns True to validate the check made in `GenerationMixin.generate()`."""
         return True
+
+    class FastTopKLogitsWarper(LogitsWarper):
+        r"""Return [batch_size, top_k] scores and indices instead of [batch_size, vocab_size] scores"""
+
+        def __init__(self, top_k: int, filter_value: float):
+            self.top_k = top_k
+            self.filter_value = filter_value
+
+        def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
+            top_k = min(self.top_k, scores.size(-1))  # Safety check
+            # Remove all tokens with a probability less than the last token of the top-k
+            return torch.topk(scores, top_k)
+
+    # Adapted from transformers.generation.utils.GenerationMixin.sample
+    def sample(
+        self,
+        input_ids: torch.LongTensor,
+        logits_processor: Optional[LogitsProcessorList] = None,
+        stopping_criteria: Optional[StoppingCriteriaList] = None,
+        logits_warper: Optional[LogitsProcessorList] = None,
+        max_length: Optional[int] = None,
+        pad_token_id: Optional[int] = None,
+        eos_token_id: Optional[Union[int, List[int]]] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        output_scores: Optional[bool] = None,
+        return_dict_in_generate: Optional[bool] = None,
+        synced_gpus: bool = False,
+        streamer: Optional["BaseStreamer"] = None,
+        **model_kwargs,
+    ) -> Union[SampleDecoderOnlyOutput, torch.LongTensor]:
```

Comment:
Might it be reusable by encoder-decoder architecture? If so I think it would be better to create a `NeuronGenerationMixin` and IMO it should also boost the performance of other module in Optimum, if so let's put it in optimum main repo @fxmarty

Reply:
There is already a `NeuronGenerationMixin`, but for training, and with a different implementation (it completely overloads `generate()` for a start). 

Regarding the applicability of that optimization to other backends, I did some tests on another ec2 instance that uses intel CPUs, and the results are not so convincing (the inf2 instances use AMD CPUs).